jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java

```diff
@@ -1005,4 +996,107 @@ public static URL toURL(File file) throws MalformedURLException
     {
         return file.toURI().toURL();
     }
+
+    /**
+     * Factory to create new Resource instance from a reference.
+     */
+    public interface Factory
+    {
+        /**
+         * Create a new Resource from the factory's point of view.
+         * <p>
+         * This is different then {@link ResourceFactory} in that
+         * it must return a {@link Resource} or throw an IOException,
+         * never null.
+         * </p>
+         *
+         * @param reference the string reference.
+         * @return the Resource instance
+         * @throws IOException if unable to create a Resource reference
+         */
+        Resource newResource(String reference) throws IOException;
```

Comment:
Do we really need a new public interface just because `ResourceFactory` can return null?    It's only called in a few spots in the methods below and null checking there would be better than introducing a nearly identical interface..... or just use `Function<String, Resource>`

We could perhaps change `ResourceFactory` in 10 to throw, but that is a lot more expensive for 404s

Reply:
The pre-existing `ResourceFactory` is really a `ResourceAddPathFactory` or `SubResourceFactory`, it's badly named.

I would love to use something like `Function<String,Resource>` but that doesn't allow for throwing of the exception.

Reply:
If we can skip the `Context.newResource(String)` layer for extraClassPath parsing then the `Resource.Factory` can be removed.

Reply:
This layer has been removed in this PR for `jetty-9.4.x`.
It should show up again in a new PR for `jetty-10.0.x`